src/onnx/parse_einsum.cpp

```diff
@@ -0,0 +1,766 @@
+                                   -1,
+                                   static_cast<ssize_t>(calc_dim(sum_axes, op2_lens))};
+
+        op1 = info.add_instruction(make_op("reshape", {{"dims", dims1}}), op1);
+        op2 = info.add_instruction(make_op("reshape", {{"dims", dims2}}), op2);
+        op2 = info.add_instruction(make_op("transpose", {{"permutation", {0, 2, 1}}}), op2);
+        instruction_ref op = info.add_instruction(make_op("dot"), op1, op2);
+
+        std::vector<size_t> new_lens(op1_lens.size(), 1);
+        for(auto i = 0; i < new_lens.size() - sum_axes.size(); ++i)
+            new_lens[i] = std::max(op1_lens[i], op2_lens[i]);
+
+        op = info.add_instruction(make_op("reshape", {{"dims", new_lens}}), op);
+
+        // compute output row
+        std::transform(cur_pair[0].begin(),
+                       cur_pair[0].end(),
+                       cur_pair[1].begin(),
+                       cur_pair[1].begin(),
+                       [](int lhs, int rhs) { return std::max(lhs, rhs); });
+        for(int a : sum_axes)
+            cur_pair[1][a] = -1;
+
+        return op;
+    }
+
+    instruction_ref finalize_output(const onnx_parser::node_info& info,
+                                    instruction_ref op,
+                                    const int_mat& map_mat,
+                                    int_mat& cur_pair) const
+    {
+        if(any_of(map_mat.back(), [](auto i) { return i >= 0; }))
+        {
+            cur_pair[1] = map_mat.back();
+            int_vec red;
+            for(int d = 0; d < map_mat[0].size(); ++d)
+            {
+                if(cur_pair[0][d] > 0 and cur_pair[1][d] == -1)
+                    red.push_back(d);
+            }
+
+            op = apply_reduce_sum_op(info, op, red, cur_pair[1]);
+        }
+
+        return transpose_squeeze(info, cur_pair, op, map_mat.back());
+    }
+
+    instruction_ref unsqueeze_transpose(const onnx_parser::node_info& info,
+                                        int_mat& cur_pair,
+                                        instruction_ref op) const
+    {
+        int_vec unsq_axes;
+        std::vector<std::tuple<int, int>> perm;
+
+        for(auto i = 0; i < cur_pair[1].size(); ++i)
+        {
+            if(cur_pair[1][i] == -1)
+                unsq_axes.push_back(i);
+            else
+                perm.push_back({cur_pair[1][i], i});
+        }
+        op = info.add_instruction(make_op("unsqueeze", {{"axes", unsq_axes}}), op);
+
+        std::sort(perm.begin(), perm.end(), [](auto lhs, auto rhs) {
+            return std::get<0>(lhs) < std::get<0>(rhs);
+        });
+
+        int_vec new_perm(cur_pair[1].size());
+        std::iota(new_perm.begin(), new_perm.end(), 0);
+
+        for(auto i = 0, p = 0; i < cur_pair[1].size(); ++i)
+        {
+            if(cur_pair[1][i] == -1)
+                continue;
+
+            new_perm[std::get<1>(perm[p++])] = i;
```

Comment:
So `std::get<1>(perm[p++])` should be the same value as `i`. What is this trying to do?

Reply:
Since `perm` is sorted before this, it won't be the same value.
The idea behind this entire `unsqueeze_transpose` function is to expand the input dimensions to match the number of unique labels in the entire equation. Then we permute the labels so they are in alphabetical order, which is what this part of the code does. This makes the logic afterward significantly easier.

Reply:
It seems like you could do something like `auto new_perm = sort_permutation(perm, by(std::less<>{}, [](auto x) { return x; })`(or maybe `std::greater<>{}` needs to be used). And then you can extend the vector with `iota` to add the missing axes.

Reply:
It also would be good to have a comment explaining what this function is trying to do as well.

Reply:
I added some comments, hope they are helpful. Regarding `sort_permutation`, it is tricky to make use of it here since the unsqueezed dims have to remain at their original axis while the other dims have to be sorted. The unsqueezed dims represent labels missing from the input term and can appear anywhere, we can't just `iota` them at the end.

e.g. `2 -1 0 -1 1` has to become `0 -1 1 -1 2` where `-1` is an unsqueezed dim, so the permutation produced needs to be `4 1 0 3 2`.

Reply:
So why not insert a transpose first before the unsqueeze? Then you dont need to compute the permutation after the unsqueeze.

Reply:
Also, if you must do the transpose after the unsqueeze(which seems more complicated), you could add a `sort_if` function to do the sorting but skip over some elements(this will at least make the intention more clear):

```cpp
template<class Iterator, class Predicate>
void sort_if(Iterator first, Iterator last, Predicate pred) {
    std::vector<typename std::iterator_traits<Iterator>::value_type> temp;
    std::copy_if(first, last, std::back_inserter(temp), pred);
    std::sort(temp.begin(), temp.end());
    auto temp_start = temp.begin();
    std::for_each(first, last, [&](auto& elem) {
        if (pred(elem))
        {
            elem = std::move(*temp_start);
            temp_start++;
        }
    });
}
```

Reply:
I refactored both `unsqueeze_transpose` and `transpose_squeeze` and renamed them to reflect the operation order (to `transpose_unsqueeze` and `squeeze_transpose`). Let me know if there's anything else that needs to be addressed.